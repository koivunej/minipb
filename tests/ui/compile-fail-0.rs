use minipb::{Status, DecodingError, Reader, io_ext::read::ReadWrapper};

use std::borrow::Cow;

/// Making sure that Reader<'static> cannot be passed over to ReadWrapper. If it could be, the
/// returned lifetime would not be valid after mutation but luckily the read_next() binding `'a` to
/// `&mut self` stops it.
struct BadReader;

impl Reader<'static> for BadReader {
    type Returned = Cow<'static, [u8]>;

    fn next(&mut self, buf: &mut &'static [u8]) -> Result<Result<Self::Returned, Status>, DecodingError> {
        Ok(Ok(Cow::Borrowed(buf)))
    }
}

fn test(read: impl std::io::Read + 'static) {
    let mut rw = ReadWrapper::new(read, BadReader);
    rw.read_next();
}

fn main() {}
