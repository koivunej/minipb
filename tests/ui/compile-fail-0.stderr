error[E0597]: `rw` does not live long enough
  --> $DIR/compile-fail-0.rs:20:5
   |
20 |     rw.read_next();
   |     ^^------------
   |     |
   |     borrowed value does not live long enough
   |     argument requires that `rw` is borrowed for `'static`
21 | }
   | - `rw` dropped here while still borrowed
